NOISSUE - RabbitMQ build and deployment

rodneyosodo opened the pull request:
### What does this do?
It aggregates the different message broker deployments starting with rabbitmq

### Which issue(s) does this PR fix/relate to?
Resolves #1610

### List any changes that modify/break current functionality
We are modularising the backend broker from nats to either nats or rabbitmq or Kafka

### Have you included tests for your changes?
No

### Did you document any new/modified functionality?
Yes

### Notes
N/A

manuio committed:
MF-1263 - Move repeating errors to the separate package (#1540)

* MF-1263 - Mv duplicated errors to pkg/errors

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Revert test build flags

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix merge

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix comment

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
NOISSUE - Fix auth members list response (#1555)

* NOISSUE - Fix auth members list response

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Move group type next to page details

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Rm membersRes

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix typo

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
MF-1261 - Use StatusUnauthorized for authn and StatusForbidden for authz (#1538)

* MF-1261 - Use StatusUnauthorized for authn and StatusForbidden for authz

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* ErrExternalKey typo

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Rename ErrUnauthorizedAcces -> ErrAuthentication

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix bootstrap error

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix status code in openapi

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix test description

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix test description

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix test description

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Add errors cases

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix status codes

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Add gRPC stutus code

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix tests description

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix openapi and encodeError

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix grpc message

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix test descriptions

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Revert sdk error

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix typo

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

blokovi committed:
MF-1059 - Add TLS support for email (#1560)

* Use gomail package for sending emails

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* remove print err

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* Add vendor

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* Rename email structure
remove logger

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* typo in var name

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* rename var

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* remove MF_EMAIL_SECRET

Signed-off-by: Ivan Milosevic <iva@blokovi.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit of adding rabbitmq broker

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit of adding rabbitmq broker

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit of adding rabbitmq broker

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit of adding rabbitmq broker

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit of adding rabbitmq broker

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit of adding rabbitmq broker

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit of adding rabbitmq broker

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

tzzed committed:
NOISSUE - Refactor MQTT subscriber (#1561)

* correct suscriber interface validator + refactore token error handling

Signed-off-by: tzzed <zerouali.t@gmail.com>

* apply review suggestion

Signed-off-by: tzzed <zerouali.t@gmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit for tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit for tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit for tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit for tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

mteodor committed:
MF-1257 - Access messages from readers endpoint with user access token (#1470)

* remove owner id

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

* add user auth for db reader

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* add user auth for db reader

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* enable mongodb reader for user token reading

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* use uuid check for auth switch between thing key and user tok

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* enable user token reading

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* revert to correct version

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fix endpoint test, add additional tests

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* remove logs,dead code

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fix logging messages

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* remove auth interface, add authorization header type

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* update api doc

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* remove unused package

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* some refactor of cases for authorization switch

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* correct description in openapi

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fix endpoint test to match auth service change

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* some rename

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* initialize auth url

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* add env variables for auth service

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fix spelling

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* Things prefix and no prefix for Thing authorization, Bearer for user

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* update readme file

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fix default things grpc port

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* enable user reading for timescaledb

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* remove not used error

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* improve errors

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* refactor authorize

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* add chanID check

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* inline some error checking

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fixing errors

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fixing errors

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* improve test case description

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* remove test code

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* dont inline

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* refactor a bit encodeError

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* remove unused error

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* remove unused error

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* fix things auth grpc url

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

* rename variables for header prefix

Signed-off-by: mteodor <mirko.teodorovic@gmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Bump up tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Bump up tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Bump up tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Bump up tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Add more tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Add more tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Add more tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Add more tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Add go routines

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Add go routines

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Add go routines

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Add go routines

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'MF-1416' of https://github.com/0x6f736f646f/mainflux into MF-1416

manuio committed:
MF-1551 - Fix Cobra usage commands and clean unnecessary struct types (#1558)

* MF-1551 - Fix Cobra usage commands and clean unnecessary struct types

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Use linux syntax for cmd usage description

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix typo

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix cmd.Use

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

blokovi committed:
NOISSUE - Separate Keto hosts for read and write (#1563)

* Separate keto hosts for read and write

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* update readme with new envars

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

* rename read connection name

Signed-off-by: Ivan Milosevic <iva@blokovi.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Co-authored-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Fix tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Fix tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
Update dependencies (#1564)

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Fix with wait groups

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Fix with wait groups

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
unsubscribe to stop delivering messages

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
unsubscribe to stop delivering messages

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
MF-1240 - Return to service transport layer only service errors (#1559)

* MF-1240 - Return to service transport layer only service errors

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Remove unecessary errors

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Rm duplicated errors and fix transport

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Revert http endpoint_test

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix conflict

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Remove exclusivity

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Remove exclusivity

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'master' into MF-1416

rodneyosodo committed:
Implement cancel mechanisms

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Implement cancel mechanisms

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

bioinformatx committed:
MF-1469 - Indicate proper authentication scheme in Authorization header (#1523)

* MF-1469 - Indicate proper authentication scheme in Authorization header

Signed-off-by: Stefan Kovacevic <jen2tri@gmail.com>

* Fixing the remarks on the last push

Signed-off-by: Stefan Kovacevic <jen2tri@gmail.com>

* Remove Bearer prefix in all services and fix tests

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix remarks

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

Co-authored-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Queuename as parameter

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Queuename as parameter

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Queuename as parameter

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Queuename as parameter

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
NOISSUE - Add nats wrapper for COAP (#1569)

* Add nats wrapper for COAP

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Pass pubsub as argument

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Defer close connection

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Defer close connection

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Rename endpoint to topic

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'master' into MF-1416

manuio committed:
MF-1348 - Add transport errors logging (#1544)

* MF-1348 - Add go-kit transport level logging

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix reviews

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix reviews

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix merge

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix remark

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix go test flags

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Use httputil errors in things and http service

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix SDK tests

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Use httputil errors in certs and provision service

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Use httputil errors in consumers service

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* General renaming and add ErrMissingToken

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Rename httputil -> apiutil and use errors in users servive

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Use apiutil errors in auth, bootstrap, readers, things and twins

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Replace errors.Contain by comparison

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix remarks

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Simplify validateID

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Simplify validateID

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Simplify and rename ExtractAuthToken -> ExtractBearerToken

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix readers

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix auth key test and remarks

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Improve comment

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Simplify validateUUID check

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix typo

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Initial commit

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'master' into RabbitMQ-deployment

dborovcanin commented on docker/docker-compose.yml:
```diff
@@ -305,11 +321,12 @@ services:
     depends_on:
       - things
       - nats
+      - rabbitmq
```
Adding both Nats and RabbitMQ as dependencies is suboptimal. The one should choose between them, rather than always having them running together. Consider using multiple compose files (one for broker and the other for the Mainflux services), updating Makefile, or checking if there are options to do it with a single compose file.

dborovcanin reviewed (changes requested):
New `messaging/broker` and another wrapper look unnecessary.

rodneyosodo committed:
Remove broker type

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
MF-1567 - Use Bearer, Thing or Basic scheme in Authorization header (#1568)

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
MF-1565 - Document Bearer, Thing and Basic Authorization header (#1566)

* MF-1565 - Document Bearer Authorization header

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix auth, bootstrap, http and readers openapi

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix openapi

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Add enc key for bootstrap

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix typo

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Use global security

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix bearer formats

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Polish descriptions

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Fix boostrap and typo

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

Co-authored-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Configure broker with makefile

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
Merge branch 'master' into RabbitMQ-deployment

rodneyosodo committed:
Check if we have changed MESSAGE_BROKER_FILE~

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'RabbitMQ-deployment' of https://github.com/0x6f736f646f/mainflux into RabbitMQ-deployment

rodneyosodo committed:
Set nats automatically

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Catching null and invalid broker type

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Remove unused fmt

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

ibalboteo committed:
MF-1575 Add 'Name' field to ListMembers response in things svc (#1576)

Signed-off-by: Ivan Balboteo <ivan@submer.com>

Co-authored-by: Ivan Balboteo <ivan@submer.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'master' of https://github.com/mainflux/mainflux into MF-1416

rodneyosodo committed:
Merge branch 'master' into RabbitMQ-deployment

rodneyosodo committed:
Merge branch 'mainflux:master' into RabbitMQ-deployment

rodneyosodo committed:
Merge branch 'RabbitMQ-deployment' of https://github.com/0x6f736f646f/mainflux into RabbitMQ-deployment

Hasan98-git committed:
MF-1580 - Influxdb Writer changes format of update-time to string (#1581)

* - MF-1580
- Modified consumers/writers/influxdb/fields.go
- influxdb-writer used to update data type of update-time to string
- Commented line 12 of consumers/writers/influxdb/fields.go to resolve
  uneccessary data type conversion issue

Signed-off-by: Hasan Tariq <hasantariqashraf@gmail.com>

* - MF-1580
- Removed strconv package from consumers/writers/influxdb/fields.go since it is no longer needed
- Removed line 12 from consumers/writers/influxdb/fields.go
- Replaced retrun value of updateTime with msg.UpdateTime (line 16 in
  fields.go)

Signed-off-by: Hasan Tariq <hasantariqashraf@gmail.com>

* Fix InflxuDB readers

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

Co-authored-by: Hasan Tariq <hasant@plcgroup.com>
Co-authored-by: dusanb94 <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
NOISSUE - Unify MF_INFLUX_READER_DB_HOST and MF_INFLUX_WRITER_DB_HOST envars (#1585)

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
NOISSUE - Fix CoAP adapter (#1572)

* Revert "NOISSUE - Add nats wrapper for COAP (#1569)"

This reverts commit cc5d5195ab27fa94270ada616487b7053fd9c7bd.

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Fix CoAP adapter

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Update CoAP observation cancel

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Fix observe

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Fix GET handling

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Revert authorization

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Use constants instead of magic numbers

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Remove an empty line

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Extract special observe value to constant

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
MF-1582 - Fix lora-adapter MQTT client (#1583)

* MF-1582 - Fix lora-adapter MQTT clien

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Add timeout config to the mqtt subscriber

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Rm comment

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* Add sub timeout

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
NOISSUE - Update changelog and readme for release 0.13.0 (#1592)

* Update release example

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Update changelog and examples for 0.13.0 release

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
Update VerneMQ release (#1593)

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
NOISSUE - Update changelog for release 0.13.0 (#1595)

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'master' of https://github.com/mainflux/mainflux into MF-1416

rodneyosodo committed:
Merge branch 'master' into RabbitMQ-deployment

rodneyosodo committed:
Fix tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Nats

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
unexport constants

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
unexport constants

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change routingkey

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change routingkey

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Remove wait groups

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Remove wait groups

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
protecting map

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
protecting map

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Add publisher to pubsub

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Add publisher to pubsub

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change proto library

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change proto library

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Fix typos

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Fix typos

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Reduce pubsub tests based on implementation

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Reduce pubsub tests based on implementation

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Remove channel cancel

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Remove channel cancel

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Export constant

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Export constant

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

fuzhy committed:
NOISSUE - Move invariant statements out of loop for cassandra-writer (#1596)

Signed-off-by: fuzhy <fuzhy1997@outlook.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
Merge branch 'master' into MF-1416

rodneyosodo committed:
Embedding publisher into pubsub

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Embedding publisher into pubsub

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio commented on cmd/mqtt/main.go:
```diff
@@ -62,9 +62,9 @@ const (
-	// Nats
-	defNatsURL = "nats://localhost:4222"
-	envNatsURL = "MF_NATS_URL"
+	// Message broker
```
Do we need this and other comments? Looks like it's the only service where it's done.

rodneyosodo replied:
This was initially done but I am changing it.

dborovcanin committed:
NOISSUE - Fix Nginx entrypoint script (#1597)

* Fix Nginx entrypoint script

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Update dependencies

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Fix NginX entrypoint

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Revert Makefile changes

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

drasko commented on Makefile:
```diff
@@ -111,4 +111,11 @@ rundev:
+else
+	echo "Invalid broker type"; exit 1
+endif
+	docker-compose -f docker/docker-compose.yml up -d
```
Remove `-d`

drasko replied:
@0x6f736f646f please take care of these remarks

arvindh123 committed:
MF-1525 - Add graceful stop for HTTP and GRPC servers (#1548)

* Add : errgroup to cmd/auth

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add : Handle graceful stop for auth service
Remove : errgroups from auth service

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add : Wait till server shutdown

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Change : instead of waitgroup changed to errgroups

Signed-off-by: Arvindh <arvindh91@gmail.com>

* change : KillSignalHandler return type to error

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Empty Commit

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add : Context to http server shutdown
Rename : varaible from proto to protocol

Signed-off-by: Arvindh <arvindh91@gmail.com>

* change : to default log level

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add : Sign-off

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add: graceful stop of http and grpc server

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Fix: typos and caps

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add: Signed-off

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Rename: Func KillSignalHandler to SignalHandler
Add: SIGABRT

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Fix: auth service

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add: timeout for grpc gracefulstop
Fix: typos

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add: .vscode folder to git ignore

Signed-off-by: Arvindh <arvindh91@gmail.com>

* change: variable name to stopWaitTime

Signed-off-by: Arvindh <arvindh91@gmail.com>

* remove: .vscode folder

Signed-off-by: Arvindh <arvindh91@gmail.com>

* remove: .vscode from .gitignore

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add : logger to handlers

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Add : New line at end of .gitignore file

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Fix : variable naming
Add : graceful stop for timescale

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Remove : unsued NATS library from import

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Move: "https" and "https" to moved to const var

Signed-off-by: Arvindh <arvindh91@gmail.com>

* Move: "http" and "https" to moved to const var

Signed-off-by: Arvindh <arvindh91@gmail.com>

* update:  branch with master

Signed-off-by: Arvindh <arvindh91@gmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Co-authored-by: Drasko DRASKOVIC <drasko.draskovic@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Naming publisher

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Naming publisher

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
MF-1588 - Update Subscriber interface (#1598)

* Initial commit

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Update subscriber interface

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

* Add tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Add tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* check subscription map

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Check topic id after topic

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* reword description

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Setup empty queue

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Change mqtt implementation

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Switch statements

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Simplify

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Change mqtt subscriber

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Protect subscription map

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Fix subscription

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Set client id

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Format

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Change delete method

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'master' into MF-1416

rodneyosodo committed:
Update rabbitmq subscriber interface

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Update rabbitmq subscriber interface

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
MF - 1590 - Fix fetching list of users with a zero limit (#1594)

* Add max and min limit size

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Format

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

* Format

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

Co-authored-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
using publisher composition

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
using publisher composition

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change contenttype

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change contenttype

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
rename topic for publish and subscribe

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
rename topic for publish and subscribe

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

manuio committed:
NOISSUE - Retrieve client key on cert issuing (#1607)

Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

zhang-chuanfeng committed:
fix bug (#1604)

Signed-off-by: zhangchuanfeng <654300242@qq.com>
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change errors to lower case

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change errors to lower case

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change errors to lower case

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change errors to lower case

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'master' into MF-1416

rodneyosodo committed:
export errors

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
export errors

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'MF-1416' of https://github.com/0x6f736f646f/mainflux into MF-1416

rodneyosodo committed:
queue per subscription

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
queue per subscription

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
queue per subscription

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
queue per subscription

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'master' into RabbitMQ-deployment

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'MF-1416' into RabbitMQ-deployment

rodneyosodo committed:
Merge branch 'master' into RabbitMQ-deployment

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
rename nats to broker

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change routing method

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change routing method

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'MF-1416' into RabbitMQ-deployment

rodneyosodo committed:
Direct method with one exchange to many queues, one consumer per queue

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Direct method with one exchange to many queues, one consumer per queue

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'MF-1416' into RabbitMQ-deployment

drasko commented on pkg/messaging/broker/pubsub.go:
```diff
@@ -0,0 +1,42 @@
+)
+
+const (
+	// chansPrefix = "channels"
```
Dead code?

drasko commented on pkg/messaging/broker/pubsub.go:
```diff
@@ -0,0 +1,42 @@
+
+// NewPubSub This aggregates the NewPubSub function for all message brokers
+func NewPubSub(url, queue string, logger log.Logger) (nats.PubSub, error) {
+	brokerSelection := mainflux.Env(envBrokerType, defBrokerType)
```
I would prefer calling this one `brokerType` and not `brokerSelection`

drasko commented on pkg/messaging/rabbitmq/publisher.go:
```diff
@@ -0,0 +1,84 @@
+	err = pub.ch.Publish(
+		exchangeName,
+		subject,
+		false,
```
@dusanb94 maybe fields specific to different brokers should be grouped in `Options` substructure?

dborovcanin replied:
What exactly would be this structure? A message field? Or like the struct you pass when creating a new Pub(Sub) instance, like configuration?

rodneyosodo committed:
:recycle: Not casting data

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
:pencil2: Fix typo

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
:recycle: remove passed queue name

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
:fire: removing echange kind

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Combine tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Refactor unsubscribe method

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'master' of https://github.com/0x6f736f646f/mainflux into MF-1416

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Fix merge conflict

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
:white_check_mark: sub and unsub to dummy topic

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin commented on pkg/messaging/broker/pubsub.go:
```diff
@@ -0,0 +1,42 @@
+type PubSub nats.PubSub
+
+// NewPubSub This aggregates the NewPubSub function for all message brokers
+func NewPubSub(url, queue string, logger log.Logger) (nats.PubSub, error) {
```
You can join these files (`publisher.go` and `pubsub.go`) to a single `pkg/messaging/brokers/brokers.go`.

rodneyosodo replied:
So with build tags, we need actually different files for Kafka, nats and rabbitmq

rodneyosodo replied:
Sorry this is in the Kafka branch

dborovcanin commented on pkg/messaging/broker/publisher.go:
```diff
@@ -0,0 +1,41 @@
+
+// NewPublisher This aggregates the NewPublisher function for all message brokers
+func NewPublisher(url string) (nats.Publisher, error) {
+	brokerSelection := mainflux.Env(envBrokerType, defBrokerType)
```
It's probably better to receive broker type as a parameter and use `mainflux.Env` only in `main.go`. The reason is sneaky configuration bugs that can arise when loading config from env this way (encapsulated in the method itself).

dborovcanin commented on pkg/messaging/broker/pubsub.go:
```diff
@@ -0,0 +1,42 @@
+// NewPubSub This aggregates the NewPubSub function for all message brokers
+func NewPubSub(url, queue string, logger log.Logger) (nats.PubSub, error) {
+	brokerSelection := mainflux.Env(envBrokerType, defBrokerType)
+	if brokerSelection == "nats" {
```
Use `switch-case` instead `if-else`.

dborovcanin commented on pkg/messaging/broker/publisher.go:
```diff
@@ -0,0 +1,41 @@
+// NewPublisher This aggregates the NewPublisher function for all message brokers
+func NewPublisher(url string) (nats.Publisher, error) {
+	brokerSelection := mainflux.Env(envBrokerType, defBrokerType)
+	if brokerSelection == "nats" {
```
Use `switch-case` instead of `if-then-else`.

rodneyosodo committed:
generate client id from topic and ID

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Rename topicID to clientID

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
update tests

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Merge branch 'MF-1416' into RabbitMQ-deployment

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
merge changes

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin reviewed (commented):
Once #1562 is merged, this PR will need to rebase. Also, docs and deployment need to be updated: https://github.com/mainflux/docs and https://github.com/mainflux/devops.

rodneyosodo committed:
Merge branch 'master' into RabbitMQ-deployment

dborovcanin commented on pkg/messaging/broker/brokers.go:
```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package broker
```
Use plural - `brokers`.

dborovcanin commented on pkg/messaging/broker/brokers.go:
```diff
@@ -0,0 +1,66 @@
+const (
+	// SubjectAllChannels represents subject to subscribe for all the channels.
+	SubjectAllChannels = "channels.>"
+)
```
```go
// SubjectAllChannels represents subject to subscribe for all the channels.
const SubjectAllChannels = "channels.>"
```

dborovcanin commented on pkg/messaging/broker/brokers.go:
```diff
@@ -0,0 +1,66 @@
+
+var (
+	errEmptyBrokerType = errors.New("empty broker type")
+)
```
```go
var errEmptyBrokerType = errors.New("empty broker type")
```

dborovcanin commented on pkg/messaging/broker/brokers.go:
```diff
@@ -0,0 +1,66 @@
+	errEmptyBrokerType = errors.New("empty broker type")
+)
+
+// PubSub type
```
Create wrapper interfaces here like:
```go

// Publisher interface enriched with connection closing.
type Publisher interface {
	messaging.Publisher
	Close() error
}

// PubSub interface enriched with connection closing.
type PubSub interface {
	messaging.PubSub
	Close() error
}

// NewPublisher This aggregates the NewPublisher function for all message brokers
func NewPublisher(brokerType, url string) (Publisher, error) {
	// brokerType := mainflux.Env(envBrokerType, defBrokerType)
	switch brokerType {
	case "nats":
		pb, err := nats.NewPublisher(url)
		if err != nil {
			return nil, err
		}
		return pb, nil
	case "rabbitmq":
		pb, err := rabbitmq.NewPublisher(url)
		if err != nil {
			return nil, err
		}
		return pb, nil
	default:
		return nil, errEmptyBrokerType
	}
}

// NewPubSub This aggregates the NewPubSub function for all message brokers
func NewPubSub(brokerType, url, queue string, logger log.Logger) (PubSub, error) {
	// brokerType := mainflux.Env(envBrokerType, defBrokerType)
	switch brokerType {
	case "nats":
		pb, err := nats.NewPubSub(url, queue, logger)
		if err != nil {
			return nil, err
		}
		return pb, nil
	case "rabbitmq":
		pb, err := rabbitmq.NewPubSub(url, queue, logger)
		if err != nil {
			return nil, err
		}
		return pb, nil
	default:
		return nil, errEmptyBrokerType
	}
}

```
A small modification on broker implementations will be needed. Also, case `MQTT` is missing in this switch.

rodneyosodo replied:
is MQTT used as an internal message broker or an adapter to provide an MQTT API for sending messages through the platform?

dborovcanin replied:
It is primarily used for forwarding MQTT messages, but It can be used for both.

rodneyosodo replied:
Should we modify the current MQTT adapter to allow for it to be used as an option in the internal broker?

dborovcanin replied:
No need to modify adapter, only the messages Pub/Sub.

rodneyosodo committed:
:recycle: rename constant and variable declaration

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin commented on pkg/messaging/broker/brokers.go:
```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package brokers
```
/pkg/messaging/**brokers**/brokers.go

dborovcanin reviewed (changes requested):
@0x6f736f646f Please be careful and at least do basic compile tests. This code won't compile due to a missing import and an error in package naming. When you fix that, you'll need to fix other broker implementations to match the `Close` method signature.

rodneyosodo committed:
rename to brokers

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
:recycle: Rename brokers

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Change mqtt implementation

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Rename broker to brokers

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
include nats port

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
move to build tags

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
move to build tags

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
add nats build tag to test

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
Merge branch 'master' into RabbitMQ-deployment

dborovcanin commented on Makefile:
```diff
@@ -111,4 +117,11 @@ rundev:
+else
+	echo "Invalid broker type"; exit 1
+endif
+	docker-compose -f docker/docker-compose.yml up
```
Add a new line at the end of the Makefile.

dborovcanin replied:
This one is still not resolved.

dborovcanin commented on pkg/messaging/brokers/brokers_nats.go:
```diff
@@ -0,0 +1,37 @@
+//go:build nats
+// +build nats
```
Since Nats is the default, consider changing logic to:
```go
//go:build !rabbitmq
// +build !rabbitmq
```
This way we do not have to provide a `nats` tag to build with nats - it will work without tags and we have a more straightforward default build.

dborovcanin commented on docker/.env:
```diff
@@ -6,8 +6,20 @@ MF_NGINX_SSL_PORT=443
-MF_NATS_URL=nats://nats:4222
+# Message Broker
+MF_BROKER_TYPE=nats
+MF_BROKER_URL=nats://broker:4222
```
The broker URL needs to change depending on the broker type - here, it's hardcoded to use NATS.

dborovcanin replied:
This is still not addressed.

dborovcanin commented on pkg/messaging/brokers/brokers_nats.go:
```diff
@@ -0,0 +1,37 @@
+//go:build nats
+// +build nats
+
```
Consider adding `init` function to log for which broker the binary is built (https://go.dev/doc/effective_go#init).

dborovcanin commented on pkg/messaging/brokers/brokers_rabbitmq.go:
```diff
@@ -0,0 +1,36 @@
+//go:build rabbitmq
+// +build rabbitmq
+
+// Copyright (c) Mainflux
```
The same comment for `init` function.

rodneyosodo committed:
\n at the end of file

Signed-off-by: GitHub <noreply@github.com>

rodneyosodo committed:
Add init function

Signed-off-by: GitHub <noreply@github.com>

rodneyosodo committed:
Add init function

Signed-off-by: GitHub <noreply@github.com>

rodneyosodo committed:
broker url automatically

Signed-off-by: GitHub <noreply@github.com>

rodneyosodo committed:
make nats as the default broker

Signed-off-by: GitHub <noreply@github.com>

rodneyosodo committed:
Merge branch 'master' into RabbitMQ-deployment

dborovcanin committed:
Merge branch 'master' into RabbitMQ-deployment

dborovcanin commented on pkg/messaging/brokers/brokers_nats.go:
```diff
@@ -0,0 +1,42 @@
+type (
+	Publisher nats.Publisher
+	PubSub    nats.PubSub
+)
```
Let's simplify remove these wrappers and update the base interface with `Close`:
```go
// Publisher specifies message publishing API.
type Publisher interface {
	// Publishes message to the stream.
	Publish(topic string, msg Message) error

	// Close gracefully closes message publisher's connection.
	Close() error
}

// MessageHandler represents Message handler for Subscriber.
type MessageHandler interface {
	// Handle handles messages passed by underlying implementation.
	Handle(msg Message) error

	// Cancel is used for cleanup during unsubscribing and it's optional.
	Cancel() error
}

// Subscriber specifies message subscription API.
type Subscriber interface {
	// Subscribe subscribes to the message stream and consumes messages.
	Subscribe(id, topic string, handler MessageHandler) error

	// Unsubscribe unsubscribes from the message stream and
	// stops consuming messages.
	Unsubscribe(id, topic string) error

	// Close gracefully closes message subscriber's connection.
	Close() error
}
```
and use `messaging.{Publisher,Subscriber,PubSub}` everwhere instead of wrapped versions (`brokers.PubSub`, `nats.PubSub`, `mqtt.PubSub`...).

rodneyosodo committed:
Updated publisher and subscriber interface

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin committed:
Merge branch 'master' into RabbitMQ-deployment

dborovcanin commented on docker/brokers/nats.yml:
```diff
@@ -0,0 +1,8 @@
+    volumes:
+      - ./../nats/:/etc/nats
+    ports:
+      - ${MF_NATS_PORT}:${MF_NATS_PORT}
```
Please add an empty line.

dborovcanin commented on docker/brokers/rabbitmq.yml:
```diff
@@ -0,0 +1,11 @@
+      RABBITMQ_DEFAULT_VHOST: ${MF_RABBITMQ_VHOST}
+    ports:
+      - ${MF_RABBITMQ_PORT}:${MF_RABBITMQ_PORT}
+      - ${MF_RABBITMQ_HTTP_PORT}:${MF_RABBITMQ_HTTP_PORT}
```
Please add an empty line.

dborovcanin commented on pkg/messaging/pubsub.go:
```diff
@@ -26,6 +29,9 @@ type Subscriber interface {
 	Unsubscribe(id, topic string) error
+
+	// Close gracefully closes message subscriber's connection.
+	Close()
```
Close should return an error. In case the underlying implementation is not returning it, return nil.

rodneyosodo committed:
Add error to close

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin commented on pkg/messaging/brokers/brokers_rabbitmq.go:
```diff
@@ -0,0 +1,36 @@
+
+package brokers
+
+import (
```
This code won't compile due to missing import.

dborovcanin replied:
Consider adding this line `MF_BROKER_TYPE=rabbitmq make http coap mqtt opcua lora` to `/scripts/ci.sh` to add compile check to CI.

dborovcanin commented on docker/brokers/rabbitmq.yml:
```diff
@@ -0,0 +1,11 @@
+services:
+  broker:
+    image: rabbitmq:3.7-management
```
Use more recent image such as `rabbitmq:3.9.20-management-alpine`.

dborovcanin commented on pkg/messaging/rabbitmq/publisher.go:
```diff
@@ -18,15 +18,8 @@ type publisher struct {
-
 // NewPublisher returns RabbitMQ message Publisher.
-func NewPublisher(url string) (Publisher, error) {
+func NewPublisher(url string) (messaging.Publisher, error) {
```
Remove `"amqp://%s"` and let the user completely define the URL, just like we do with NATS.

rodneyosodo committed:
Change image version and let user define URI

Signed-off-by: GitHub <noreply@github.com>

rodneyosodo committed:
Merge branch 'master' into RabbitMQ-deployment

dborovcanin committed:
Merge branch 'master' into RabbitMQ-deployment

rodneyosodo committed:
make broker url configurable

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
add compile check

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
rabbitmq check

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
default broker to nats

Signed-off-by: GitHub <noreply@github.com>

dborovcanin commented on scripts/ci.sh:
```diff
@@ -65,6 +65,8 @@ setup_mf() {
         fi
     done
     make -j$NPROC
+    echo "Compile check for rabbitmq..."
+    MF_BROKER_TYPE=rabbitmq make http
```
Put this before ` make -j $NPROC`.

dborovcanin commented on scripts/ci.sh:
```diff
@@ -81,6 +83,7 @@ setup() {
     setup_lint
 }
 
+
```
Remove added empty line.

dborovcanin commented on Makefile:
```diff
@@ -111,4 +117,12 @@ rundev:
-	docker-compose -f docker/docker-compose.yml up
+ifeq ("$(MF_BROKER_TYPE)", "rabbitmq")
+	sed -i "s,file: brokers/.*.yml,file: brokers/rabbitmq.yml," docker/docker-compose.yml
+	export MF_BROKER_URL=amqp://$(MF_RABBITMQ_USER):$(MF_RABBITMQ_PASS)@broker:$(MF_RABBITMQ_PORT)$(MF_RABBITMQ_VHOST) && docker-compose -f docker/docker-compose.yml up
```
This will not work unless the user has locally set `MF_RABBITMQ_*` env vars set because it will be replaced with empty values. I would suggest doing something like this instead:
```make
	sed -i "s,file: brokers/.*.yml,file: brokers/rabbitmq.yml," docker/docker-compose.yml
	sed -i "s,MF_BROKER_URL: .*,MF_BROKER_URL: $$\{MF_RABBITMQ_URL\}," docker/docker-compose.yml
```

rodneyosodo replied:
I thought we needed to build the broker URL for instance when using rabbitmq

rodneyosodo committed:
change broker url in docker compose

Signed-off-by: GitHub <noreply@github.com>

rodneyosodo committed:
Fix ci

Signed-off-by: GitHub <noreply@github.com>

dborovcanin commented on Makefile:
```diff
@@ -111,4 +117,13 @@ rundev:
-	docker-compose -f docker/docker-compose.yml up
+ifeq ("$(MF_BROKER_TYPE)", "rabbitmq")
+	sed -i "s,file: brokers/.*.yml,file: brokers/rabbitmq.yml," docker/docker-compose.yml
+	sed -i 's#{MF_BROKER_URL}#MF_RABBITMQ_URL#' docker/docker-compose.yml
```
This approach is not flexible enough and will work only when you first replace `${MF_BROKER_URL}` with `${MF_RABBITMQ_URL}`. If you want to run it with any other broker later, it won't work. Since you [changed](https://github.com/mainflux/mainflux/pull/1570/files#diff-423deb13b7c401b1a7f41ee91c77f722e11d2f317d6a66b546524e8a04cc8b03R289) `docker-compose.yml` to use `${MF_NATS_URL}`, it will not work at all. I prefer using `MF_BROKER_URL: ${MF_NATS_URL}` over `MF_BROKER_URL: ${MF_BROKER_URL}` in `docker-compose.yml`, but Makefile `sed` command needs an update.

rodneyosodo committed:
fix makefile

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

dborovcanin commented on docker/.env:
```diff
@@ -6,8 +6,21 @@ MF_NGINX_SSL_PORT=443
+MF_RABBITMQ_PASS=mainflux
+MF_RABBITMQ_COOKIE=mainflux
+MF_RABBITMQ_VHOST=/
+MF_RABBITMQ_URL=amqp://mainflux:mainflux@broker:5672/
```
You can reuse previous values to construct the URL: `MF_RABBITMQ_URL=amqp://$MF_RABBITMQ_USER:$MF_RABBITMQ_PASS@broker:$MF_RABBITMQ_PORT$MF_MF_RABBITMQ_VHOST`

dborovcanin commented on scripts/ci.sh:
```diff
@@ -65,6 +65,8 @@ setup_mf() {
     done
     make -j$NPROC
+    echo "Compile check for rabbitmq..."
+    MF_BROKER_TYPE=rabbitmq make -j $NPROC http
```
Put this before `make -j $NPROC`: https://github.com/mainflux/mainflux/pull/1570#discussion_r901527315

dborovcanin reviewed (changes requested):
We're almost done.

rodneyosodo committed:
Fix env var and ci

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Fix ci

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

rodneyosodo committed:
Fix consts

Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

drasko reviewed (approved):
LGTM
